packages/runtime/test/integration/atomic-transfer.ts

```diff
@@ -0,0 +1,431 @@
+      RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE
+    );
+
+    assert.equal(runtime.getAssetDef(assetId).creator, john.address);
+  });
+
+  it("should fail close app if payment transaction fails", () => {
+    const txGroup: ExecParams[] = [
+      {
+        type: TransactionType.CloseSSC,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        appId: appId,
+        payFlags: { totalFee: 1000 }
+      },
+      {
+        type: TransactionType.TransferAlgo,
+        sign: SignType.SecretKey,
+        fromAccount: alice.account,
+        toAccountAddr: john.address,
+        amountMicroAlgos: 6e6,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+
+    expectRuntimeError(
+      () => runtime.executeTx(txGroup),
+      RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE
+    );
+
+    syncAccounts();
+    assert.isDefined(john.getLocalState(appId, key));
+  });
+
+  it("should fail clear app if payment transaction fails", () => {
+    const txGroup: ExecParams[] = [
+      {
+        type: TransactionType.ClearSSC,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        appId: appId,
+        payFlags: { totalFee: 1000 }
+      },
+      {
+        type: TransactionType.TransferAlgo,
+        sign: SignType.SecretKey,
+        fromAccount: alice.account,
+        toAccountAddr: john.address,
+        amountMicroAlgos: 6e6,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+
+    expectRuntimeError(
+      () => runtime.executeTx(txGroup),
+      RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_BALANCE
+    );
+
+    syncAccounts();
+    assert.isDefined(john.getLocalState(appId, key));
+  });
+
+  it("should fail asset payment, and algo payment if ssc call fails", () => {
+    // close out from app
+    runtime.executeTx({
+      type: TransactionType.ClearSSC,
+      sign: SignType.SecretKey,
+      fromAccount: john.account,
+      appId: appId,
+      payFlags: { totalFee: 1000 }
+    });
+    syncAccounts();
+    const txGroup: ExecParams[] = [
+      {
+        type: TransactionType.ClearSSC,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        appId: appId,
+        payFlags: { totalFee: 1000 }
+      },
+      {
+        type: TransactionType.TransferAlgo,
+        sign: SignType.SecretKey,
+        fromAccount: alice.account,
+        toAccountAddr: john.address,
+        amountMicroAlgos: 100,
+        payFlags: { totalFee: 1000 }
+      },
+      {
+        type: TransactionType.TransferAsset,
+        sign: SignType.SecretKey,
+        fromAccount: john.account,
+        toAccountAddr: alice.account.addr,
+        amount: 10,
+        assetID: 1,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+    const initialJohnAssets = john.getAssetHolding(assetId)?.amount as number;
+    const initialAliceAssets = alice.getAssetHolding(assetId)?.amount as number;
+    const initialJohnBalance = john.balance();
```

Comment:
is `john.balance()` different from `initialBalance` here?

Reply:
yes, because according to code closeout is executed, therefore fees is deducted